Add serial frame detector

Miceuz opened the pull request:
# Breakdown of RX on Cycle Controller side
`FrameDetector` is the main component responsible for reception, it is a Finite State Machine implementing `RxListener` interface, triggred by `RxListener` events, mainly governed by `onCharacterMatch` event. 

`RxBuffer` is an abstraction covering memory buffer filled by `UART_DMA` and providing access to control of `UART_DMA`. Mainly this abstraction is extracted here so that `FrameDetector` can be moved to commons and reused in GUI, also more easily tested. 

`UART_DMA` is set up to provide character match interrupts on reception of `FRAME_MARK` char over serial. In Cycle Controller this is provided by hardware. 

`FrameDetector` asks RxBuffer to start an UART DMA process that automatically writes incoming data to a memory buffer wrapped by `RxBuffer`. `FrameDetector` also passes itself to `RxBuffer` as a `RxListener` to get rx events.  On `CharacterMatch` event `FrameDetector` decides the transition to the next state by asking `RxBuffer` for amount of received data. `RxBuffer` asks `UART_DMA` to determine that. `FrameDetector` can ask `RxBuffer` to restart the reception - this clears the buffer and restarts `UART_DMA` reception. 

Upon successful frame detection `FrameDetector` copies the RxBuffer to a separate buffer and provides that to the caller - in our particular case - to `Comms`. 

#439 Remove CRC32 implementation 
#440 uart dma update
👉#441 159 frame detector👈 **YOU ARE HERE**
#442 Add framing streams
#443 frame decoding
#444  Update comms to use UART_DMA and framing

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/respiraworks/ventilatorsoftware/441)
<!-- Reviewable:end -->

hhff committed:
Implement gui build script and unit test harness

sglow committed:
Introduces the debug serial interface

sglow committed:
Minor tweaks based on review feedback

sglow committed:
Minor changes based on feedback from review

sglow committed:
Merge pull request #351 from RespiraWorks/issue_325_motor_driver

Issue 325 motor driver

sglow committed:
Revised A/D sampling method for improved readings

sglow committed:
Minor tweaks based on review feedback

sglow committed:
Minor changes to comments

sglow committed:
Merge pull request #353 from RespiraWorks/adc_improvements

Revised A/D sampling method for improved readings

Commit:
add task github issue template

Commit:
when to create Feature request or Task

Commit:
Merge pull request #354 from ysz/feature/task

add task github issue template

jkff committed:
Deletes controller-side alarm code - alarms will be handled in the GUI

jkff committed:
Merge pull request #357 from jkff/rm-alarms: Deletes controller-side alarm code - alarms will be handled in the GUI

asmodai27 committed:
Add functionnality to units:
+= and -= operator for ArithScalar
New ModuleScalar template for averaging/applying a multiplying factor (e.g calibration of flow?) right now only for flow and volume
Define Volume = VolumetricFlow*Duration = Duration*VolumetricFlow and VolumetricFlow = Volume/Duration

jkff committed:
Merge pull request #362 from jkff/reintro-debug: (Re-)Introduces the debug serial interface

jkff committed:
Makes ./test.sh pass with clang.

This includes one bug (= instead of ==) and something that is most
likely undefined behavior but somehow worked: all the debug cmds
were declared as static variables of type "FooCmd" in their respective
sources, but then as "extern DebugCmd" (different type) in debug.cpp.

git-pr-chain: test-sh

jkff committed:
Puts command classes in separate headers from their respective abstractions.

Names all command headers with _cmd and adds include guards.

git-pr-chain: debug-cmd-headers

jkff committed:
Merge pull request #363 from RespiraWorks/jkff/test-sh: Makes ./test.sh pass with clang.

jkff committed:
Merge pull request #364 from RespiraWorks/jkff/debug-cmd-headers: Puts command classes in separate headers from their respective abstractions

jkff committed:
Cosmetic symbol renames in vars.h

git-pr-chain: debug-vars

jkff committed:
Changes debug vars to be adapters over uint32

asmodai27 committed:
adress jlebar review comments:
- merge ModuleScalar in ArithScalar
- make ArithScalar*float commute
- make Duration*Flow and Volume/Duration work with the unit() fonctions to stay independent from implementation details
- get rid of some useless parenthesis

asmodai27 committed:
Address reviews: clarify comments and improve tests coverage of float operations

sglow committed:
Added a couple handy debug vars for tracing

sglow committed:
Restored old (wrong) diameter.  Will fix later.

Fixing this requires fixing the tests which were created based
on the old diameter.  I'll tackle that in a future PR

sglow committed:
Minor changes based on feedback from review

sglow committed:
Merge pull request #377 from RespiraWorks/add_debug_vars

Added a couple handy debug vars for tracing

jkff committed:
Adds vars_test.cpp

jkff committed:
Add test for var cmd

jkff committed:
Fix warning

jkff committed:
Merge pull request #365 from RespiraWorks/jkff/debug-vars: Changes debug vars to be adapters over uint32_t and adds basic tests

twslankard committed:
build an armv7 (aka armhf?) image instead of arm64, since the RPi can… (#379)

* build an armv7 (aka armhf?) image instead of arm64, since the RPi can't run arm64

* Issue #252: use non arch-specific Docker base image

twslankard committed:
Tag the latest master GUI docker build with `latest` in addition to SHA1 to make dev's lives easier. (#380)

NOTE: we'll still need to have a conversation about how updates are supposed to work, and whether we want to create release tags.

a-vinod committed:
Cache latest gui docker image & spin up on boot (#361)

* Run GUI container on boot

TODO code to be commented out once a respiraworks/gui docker image is on dockerhub.

* Spin up GUI Docker container on boot

TODO code will be commented out once a gui docker image is pushed to docker hub so the latest gui image is cached into the pi upon creation of the raspbian image.

Need edit Dockerfile so that the gui container runs the gui executable within it upon spinning up the container

The end result should be something like: rPi boots into Openbox -> GUI docker image spun up -> gui executable within container is run

* Cache & spin up gui docker image on boot

* Added tag to respiraworks/gui image pull & autostart

* Remove SHA1 tag

sglow committed:
Several minor debug improvements

sglow committed:
Added debug variables to allow cycling without GUI

jkff committed:
Separates the request and response buffers in DebugCmd

This will make it easier to instead pass a request and
response protobuf.

git-pr-chain: debug-req-resp

martukas committed:
populated with more READMEs and moved some content from wiki

jkff committed:
Puts tracing into a class.

git-pr-chain: trace-class

martukas committed:
updated main readme and enabled pre-commit

jlebar committed:
Pass -fno-threadsafe-statics.

This lets us use `static` variables inside of functions with types that
have nontrivial initializers.  Without this, you'll get a mysterious
error of the form

  sp/hard/libnosys.a(sbrk.o): In function `_sbrk':
  sbrk.c:(.text._sbrk+0x18): undefined reference to `end'

which indicates that something is trying to call malloc.

martukas committed:
dma uart test documentation

martukas committed:
more utils docs and controller docs more truthy

martukas committed:
some notes on hardware for testing utilities

martukas committed:
just a wee little bug

jlebar committed:
Merge pull request #384 from RespiraWorks/jlebar/threadsafe-initializers

Pass -fno-threadsafe-statics.

jkff committed:
Merge pull request #382 from RespiraWorks/jkff/debug-req-resp: Separates the request and response buffers in DebugCmd

sglow committed:
Fixed bug from earlier merge which broke debug tools

Also made some minor tweaks based on review feedback

sglow committed:
Added a comment

sglow committed:
Merge pull request #381 from RespiraWorks/minor_debug_improvements

Minor debug improvements

sglow committed:
Added SetAmpAll()

sglow committed:
Added module to control pinch valves

sglow committed:
Moved watchdog timer init to after startup code.

The startup code can take a while to run and runs before the
high priority loop starts running.  The high priority loop
is where the watchdog timer is reset, so this created a condition
where the watchdog timer could time out before we start servicing
it.

sglow committed:
Make sure ints are enabled before initing steppers

The stepper motor driver uses interrupts during it's initilization
it's therefore imporant that interrupts are enabled before the
stepper motors are initilized

sglow committed:
Fixed reversed direction

sglow committed:
Removed a script that isn't useful to others

jkff committed:
Merge pull request #272 from RespiraWorks/issue_270_remove_travis: CI Chore: Remove travis from repo, closes #270

jkff committed:
Removes references to Travis

jkff committed:
Merge pull request #385 from RespiraWorks/jkff/trace-class: Puts tracing into a class.

jkff committed:
Do not reuse buffer index between request and response.

Also:
* Adds functions for testing the debug serial port to HAL
* Adds a basic test of the Debug interface (just two commands for now)

git-pr-chain: buf-index

jkff committed:
Centralizes trace management completely in the Trace class and adds tests.

git-pr-chain: trace-tests

jkff committed:
Merge pull request #389 from RespiraWorks/jkff/circleci-yml: Removes references to Travis

martukas committed:
simple fan test

martukas committed:
"wider fan ramp range, temporary hack fix to HAL init bug; updates #390"

Miceuz committed:
Add a temporary band aid for #393 to unbreak the system by disabling compiler optimizations

Miceuz committed:
Declare cmd_ptr_ a volatile pointer to volatile data to avoid compiler optimizing out busy wait in StepMotor::SenCmd. Fixes #393

paulovap committed:
[GUI] Add script to run gui on eglfs 1920 x 1080 @ 60 mode

This script run_eglfs.sh is added alongside a kms configuration
file to force egl to run in a specific resolution/refresh_rate.

The script also enable some Qt debugging information to make it
easy to troubleshoot eventual issues with graphics

Also added a small fps counter element so we can keep track of
performance degratations. It will be available only in debug
builds once we start making this distinction.

jlebar committed:
Update jlebar/pre-commit-hooks repo.

Fixes Windows, and fixes cases where the user's tempdir is a different
filesystem than their home directory.

GPC: pre-commit-update

martukas committed:
Merge pull request #398 from RespiraWorks/jlebar/pre-commit-update

Update jlebar/pre-commit-hooks repo.

jlebar committed:
Add ProtoTraits traits class.

This lets you generically access e.g. MaxSize for T, where T is one of
our protos.

Meant to be used in https://github.com/RespiraWorks/VentilatorSoftware/pull/346.

GPC: proto-traits

martukas committed:
changes in response to PR

martukas committed:
Merge pull request #383 from RespiraWorks/martin_cleanup

Improving in-repo documentation

sglow committed:
Fixed some issues with powerSTEP chip

martukas committed:
"uncommented stepper init after rebasing; updates #390"

paulovap committed:
[GUI] Add Style singleton and extract left-side-panel controls into reusable Item

On main.qml we had few hard-coded controls for "RR", "PEEP",
"PIP" and so on. This change extracts those into a reusable
element called StepCounter which allows user to increase
or decrease a value, one step at the time.

Aside from that, a singleton element Style is created
to hold all UI related configuration, such as themes,
fonts, margin sizes and so on.

sglow committed:
Minor edits based on review

sglow committed:
Merge pull request #404 from RespiraWorks/issue_378_pinch_valve

Issue 378 pinch valve

paulovap committed:
Merge pull request #406 from RespiraWorks/ppinheiro/StepCounter

[GUI] Add Style singleton and extract left-side-panel controls into r…

paulovap committed:
Merge pull request #399 from RespiraWorks/ppinheiro/fix_eglfs_perf

[GUI] Add script to run gui on eglfs 1920 x 1080 @ 60 mode

jlebar committed:
Build with -fno-exceptions.

GPC: fno-exceptions

jlebar committed:
Clean up build flags.

- Make stm32_build_flags variable, shared between the two STM32 builds.
  This causes `-e stm32-test` to run with -fstrict-volatile.

- Filter src_test out of `native` build, so `pio run -e native` now
  doesn't try to build src_test.  (This isn't a standard command,
  usually we do `pio test -e native`.)

GPC: clean-build-flags

jlebar committed:
Build with -fno-rtti.

Per this guide to disabling exceptions in embedded C++:
http://elegantinvention.com/blog/information/smaller-binary-size-with-c-on-baremetal-g/

GPC: fno-rtti

jlebar committed:
Add an implementation of __cxa_pure_virtual.

This (together with patches earlier in the queue) allows us to use pure
virtual functions, yay.

Also use them in vars.h.

GPC: pure-virtual

paulovap committed:
[GUI] Add initial implementation on Selection Mode Popup

Mode selection popup will allow us to change the ventilator
operation mode. The design and copies are still
under development and will be refined later.

The organization of the QML files was also changed for easy understanding:

* control/ folder keeps all Pure UI controls, like buttons and widgets
* modes/ folder keeps the implementation of all modes. Later this could be
broken down if it gets too complex.

Add stub modes: PressureAssistMode and SupportPressureMode for
testing the mode selection

jlebar committed:
Merge pull request #402 from RespiraWorks/jlebar/proto-traits

Add ProtoTraits traits class.

jlebar committed:
Merge pull request #410 from RespiraWorks/jlebar/clean-build-flags

Clean up build flags.

jlebar committed:
Merge pull request #411 from RespiraWorks/jlebar/fno-exceptions

Build with -fno-exceptions.

jlebar committed:
Merge pull request #412 from RespiraWorks/jlebar/fno-rtti

Build with -fno-rtti.

jlebar committed:
Merge pull request #413 from RespiraWorks/jlebar/pure-virtual

Add an implementation of __cxa_pure_virtual.

martukas committed:
"resolved problem after rebase; updates #390"

sglow committed:
Fixed stepper commands that take no data.

sglow committed:
Added a table to linearize the pinch valve output

Also replaced the startup homing to an automatic homing state machine
that can run from the high priority loop rather then the background
task.  This has two advantages:
- It allows multiple pinch valves to be homed at the same time
- It allows the valves to be disabled when the ventilator is not in use.
  That saves power and is easier on the stepper motors and their drivers

sglow committed:
Removed obsolete call to home pinch valve on startup

martukas committed:
"fixed after rebase; updates #390"

martukas committed:
"solenoid test added; updates #390"

martukas committed:
"simple stepper motor test; updates #390"

martukas committed:
"test.sh builds integration tests; updates #390"

martukas committed:
"added pinch valve test; updates #390"

martukas committed:
"added some comments to test code; updates #390"

martukas committed:
Merge pull request #420 from RespiraWorks/issue_390

Basic vertical integration tests

sglow committed:
Made flow table length agnostic

sglow committed:
Merge pull request #419 from RespiraWorks/issue_418_pinch_valve_linearize

Issue 418 pinch valve linearize

jkff committed:
Make it possible to have several tests

jkff committed:
Merge pull request #391 from RespiraWorks/jkff/buf-index: Do not reuse buffer index between request and response.

jkff committed:
Merge pull request #422 from RespiraWorks/jkff/takeover-pr-95: [Takeover of #95] Adds Unit Test Harness for GUI, closes #87

jkff committed:
Merge pull request #392 from RespiraWorks/jkff/trace-tests: Centralizes trace management completely in the Trace class and adds tests.

paulovap committed:
[GUI Tests] Simplify GUI sub-project to run tests.

So, the current implementation of tests could be a bit
cumbersome, because needed to create a new qmake project
for each test class. This PR simplifies the gui folder structure
to accommodate tests sub-project in a way to keep all tests in a
single folder.

paulovap committed:
Merge pull request #416 from RespiraWorks/ppinheiro/mode_selection_popup

[GUI] Add initial implementation on Selection Mode Popup

jlebar committed:
First pass at TV zeroing.

Uses PID to push tidal volume to zero on each breath.

The PID could use some tuning, but as a first pass, this works OK for
me.

GPC: tv-zeroing

jlebar committed:
Set default params for NO_GUI_DEV_MODE.

Rather than starting the ventilator off, set some default values for the
NO_GUI_DEV_MODE params.

To do this, we had to modify the DebugT classes to let you specify a
default value on construction.

GPC: debug-defaults

jlebar committed:
Refactor controller_debug's `debug` command.

- Instead of a toggle, you specify 'debug on' or 'debug off'.
- Declare a global DbgPrint function instead of using a function
  pointer.

GPC: refactor-debug

jlebar committed:
Sort includes in controller_debug.py.

GPC: sort-includes

jlebar committed:
controller_debug: Print help for trace.

When you run plain `trace`, print an error and also print out `trace
help`.

To accomplish this, I reformatted the do_trace docstring so there isn't
a bunch of leading whitespace.  This matches other docstrings in here.
I didn't change the content of the docstring in this change.

GPC: print-trace-help

jlebar committed:
Fix typo: seperate -> separate.

GPC: separate

jlebar committed:
Add readline to list of imports.

This makes the cmd module use readline for handling inputs, so things
like ctrl+A work.

GPC: readline

jlebar committed:
Use argparse for `trace download` and `console`.

Using argparse is less code and gives more helpful errors, but more
importantly it will allow us to parse more complicated subcommands.

GPC: argparse

jlebar committed:
Get rid of most "global" statements.

"global" is only necessary when you assign to a global variable before
you read it.  We only have one variable like this; all of the other
"global" statements are unnecessary.

GPC: globals

jlebar committed:
Get rid of crc16 global variable.

It's unnecessary; we can just use a function.

GPC: crc16-global

jlebar committed:
Add documentation for `exec` command.

GPC: exec-documentation

jlebar committed:
Tighten up trace documentation a bit.

GPC: trace-docs

Miceuz committed:
Remove CRC32 implementation as it proved to be not that useful as it complicates code reuse between GUI and Cycle Controller. Also remove legacy 16 crc implementation

Miceuz committed:
Move serial event listeners from constructor of UART_DMA to particular startTX, startRX methods; some minor interface changes, some renaming

paulovap committed:
[GUI] Add ParameterButton element

This change has the following additions:

* Updates the HeaderButtons according to the current design document[0].
* Creates a Button element to change mode parameters on different modes.
* Adds new fonts according to the current design document[0].
* Replace parameter buttons in all modes.

0 - https://xd.adobe.com/view/5816839b-fd33-41f2-5a21-4b218fb9867f-2e27/screen/e0f99113-e872-4d96-84e4-c463f61812f1/01-mainscreen-Pressure-assist-01?fullscreen

paulovap committed:
[GUI] Add parameter display element

Adding a ParameterDisplay element that will be used
to show current value for all the relevant parameters.

Also update the design of the modes to follow more closely
the new design recommendation

jlebar committed:
Merge pull request #427 from RespiraWorks/jlebar/debug-defaults

Set default params for NO_GUI_DEV_MODE.

jlebar committed:
Merge pull request #428 from RespiraWorks/jlebar/refactor-debug

Refactor controller_debug's `debug` command.

jlebar committed:
Merge pull request #429 from RespiraWorks/jlebar/sort-includes

Sort includes in controller_debug.py.

jlebar committed:
Merge pull request #430 from RespiraWorks/jlebar/print-trace-help

controller_debug: Print help for trace.

jlebar committed:
Merge pull request #431 from RespiraWorks/jlebar/separate

Fix typo: seperate -> separate.

jlebar committed:
Merge pull request #432 from RespiraWorks/jlebar/readline

Add readline to list of imports.

jlebar committed:
Merge pull request #433 from RespiraWorks/jlebar/globals

Get rid of most "global" statements.

jlebar committed:
Merge pull request #434 from RespiraWorks/jlebar/crc16-global

Get rid of crc16 global variable.

jlebar committed:
Merge pull request #436 from RespiraWorks/jlebar/exec-documentation

Add documentation for `exec` command.

jlebar committed:
Merge pull request #437 from RespiraWorks/jlebar/trace-docs

Tighten up trace documentation a bit.

jlebar committed:
Merge pull request #438 from RespiraWorks/jlebar/argparse

Use argparse for `trace download` and `console`.

jlebar committed:
Fix up docstrings.

- Make some comments into docstrings.
- Complete incomplete docstring on TraceDownload.
- Format docstrings so the first line isn't empty (PEP 257)

GPC: docstrings

paulovap committed:
[GUI] Add stub for device status information.

Adds UI for device status (clock & battery).

Clock works out-of-the-box on computers with real time clocks,
but it was not tested in a raspberry pi.

Getting information from battery still not implemented.

paulovap committed:
Merge pull request #423 from RespiraWorks/ppinheiro/gui-tests-simplify

[GUI Tests] Simplify GUI sub-project to run tests.

Miceuz committed:
Remove unexisting include directories from platformio.ini, remove stm32_test environment as it's served its purpose

Miceuz committed:
Move pure_virtual to core so it can be seen by all

Miceuz committed:
Add frame detector and acompanying RxBuffers

jlebar committed:
Add arguments to `trace start` command.

You can now do e.g. `trace start --period 10 volume flow_diff`.

GPC: trace-start-args

Miceuz committed:
Remove a stray pure_virtual

jlebar committed:
Auto-detect port device is connected to.

GPC: port-autodetect

jlebar committed:
Tweak implementation of TraceDownload.

No functional change, just more comments and a different implementation.

GPC: trace-download

jlebar committed:
Label axes and lines in graph.

- Label x axis as seconds (and scale according to trace period).
- Label individual lines according to what we're graphing, e.g.
  "Tidal volume, ml".

GPC: labels

jlebar committed:
Revert "Removed time checks from PID library."

This reverts commit be574eaddb12501e9bfff7f74b8f7210dbe554af.

The time checks were removed from PID because we were only calling PID
from the precisely-timed control loop.  We tried to update the PID every
(say) 10ms, and we were confident we *would* call every 10ms, plus or
minus a few microseconds.  But if we happened to call with a delta of
9.99ms, the PID code would skip our sample!  That's bad.

However, for flow correction, we need our PID library to be able to
handle the case when the time interval is not constant.  In this case,
we *do* want time checks.

So in this change we add time checks back.  The next commit makes time
checking optional, so the PID library can work for both cases.

GPC: pid-time

jlebar committed:
Use /usr/bin/env python3 as shebang.

This way we pull whatever is the user's default python3, instead of
forcing /usr/bin/python3.

GPC: shebang

jlebar committed:
Add -c flag.

Lets you run a single command, e.g.

  controller_debug.py -c "trace start volume" && sleep 10 &&
  controller_debug.py -c "trace graph"

Currently, there's no way to do multiple commands except by multiple
invocations of the script.

GPC: dash-c

jlebar committed:
Add debug var for tidal volume.

GPC: tv-debug-var

jlebar committed:
Add two sample 120s volume traces.

One wiggles all around, the other is a straight line.  Why, I cannot
say.

GPC: volume-traces

jlebar committed:
Merge pull request #445 from RespiraWorks/jlebar/docstrings

Fix up docstrings.

jlebar committed:
Merge pull request #446 from RespiraWorks/jlebar/trace-start-args

Add arguments to `trace start` command.

jlebar committed:
Merge pull request #447 from RespiraWorks/jlebar/port-autodetect

Auto-detect port device is connected to.

jlebar committed:
Merge pull request #448 from RespiraWorks/jlebar/trace-download

Tweak implementation of TraceDownload.

jlebar committed:
Merge pull request #449 from RespiraWorks/jlebar/labels

Label axes and lines in graph.

jlebar committed:
Merge pull request #451 from RespiraWorks/jlebar/shebang

Use /usr/bin/env python3 as shebang.

jlebar committed:
Merge pull request #452 from RespiraWorks/jlebar/dash-c

Add -c flag.

jlebar committed:
Merge pull request #453 from RespiraWorks/jlebar/tv-debug-var

Add debug var for tidal volume.

jlebar committed:
Merge pull request #455 from RespiraWorks/jlebar/volume-traces

Add two sample 120s volume traces.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,129 @@
+  enum class State {
+    LOST,       // Frame detector does not know where it is in the stream of
+                // bytes coming into RxBuffer
+    WAIT_START, // Frame detector for a frame start marker
```
Frame detector **is waiting** for a frame start marker?

Also how about renaming the states: WAIT_START -> WAITING_FOR_START_MARKER, RX_FRAME -> RECEIVING_FRAME?

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,129 @@
+  void onCharacterMatch() override {
+    switch (state_) {
+    case State::LOST:
+      // if we have received something before this marker,
```
Is this a normal situation, or an error, or something that can only happen after an error?

Miceuz replied:
We are LOST, this means we are trying to recover from an error or we just started and trying to orient our selves in the incoming stream of data

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,129 @@
+      // for start
+      if (rx_buffer_.ReceivedLength() > 1) {
+        state_ = State::WAIT_START;
+        // if we were lucky to get lost in the interframe silence,
```
Is this comment referring to line 44 or to line 47? If to line 44, please merge with the comment above and condense; if to line 47, then please put before line 48.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,129 @@
+#include <string.h>
+
+template <class RxBuffer, int FRAME_BUF_LEN>
+class FrameDetector : public RxListener {
```
Please add some high-level comments as to how this works, e.g. a couple of typical situations explained in terms of what state transitions they cause.

Miceuz replied:
While writing these I have realized there might be some smelly stuff going on when the full frame is received - we will have a race condition between `OnCharacterMatch` and `OnRxComplete` callbacks. I will test against that and fix it if something comes up.

Miceuz replied:
Ok, we are good in the all good case. I was wrong assuming that both `OnCharacterMatch` and `OnRxComplete` will come at the same time. We will always have one char place extra in the buffer when the max size frame comes, because we discard the frame start marker (later we also discard end frame marker in ProcessReceivedData()).

Though, in case sender sends a frame that's 1 byte longer than max frame length, we *will* have this race condition. If  `OnCharacterMatch` wins we will have frame detected and end up in LOST state. If `OnRxComplete` wins we will have no frame detected and end up in `WAIT_FOR_START_MARKER` state. In both cases we are able to recover, besides sender was already wrong in sending a frame too long. We can resolve this race condition by setting hardware interrupt priorities. And in soft implementation of RxBuffer we can resolve this by explicitly coding the order in which callbacks are called like it is done now in `SoftRxBuffer`.

Miceuz replied:
After putting more thinking to it, I think I should not be discarding the start marker, so that the whole thing is not driven by side effects. I am afraid that will have implications on code clarity, but still...

Miceuz replied:
I have refactored the FrameDetector to be more explicit in handling markers. Now we don't have the bug with illegaly long frame being detected. Also I have added a testing infrastructure for fuzzing as you have requested.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,129 @@
+        state_ = State::WAIT_START;
+      } else {
+        // repeated marker char received
+        // assume we are still good
```
What does "still good" mean?

Miceuz replied:
We have received an extra frame marker just after receiving a marker we have decided to start a frame on.
```
bbbMMMffff...
    ^ go to RECEIVING_FRAME
     ^ this is an extra marker we received
       ^ assume we will receive a proper frame after this     
```

Miceuz replied:
What we can do here:
 * Just decode 0 length frame - does not make much sense, but maybe sender got crazy and is sending 0 length frames
 * go to LOST - maybe
 * ignore and continue decoding, it this MARK was caused by a bit flip in stream, upper level decoder will catch that

I will add more tests and see if it creates any problems.

Miceuz replied:
So far I think this implementation is correct as it allows us to recover in many situations related to callbacks racing.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,129 @@
+  uint8_t frame_buf_[FRAME_BUF_LEN];
+  uint32_t frame_buf_length_ = 0;
+
+  void processReceivedData() {
```
ProcessReceivedData

jkff replied:
Please rename

Miceuz replied:
Done

jkff commented on common/libs/framing/soft_rx_buffer.h:
```diff
@@ -0,0 +1,46 @@
+#ifndef __SOFT_RX_BUFFER_H
```
The include guard should be SOFT_RX_BUFFER_H_

jkff replied:
Please address this and other previous comments in this file.

Miceuz replied:
Done

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,129 @@
+#ifndef __FRAME_DETECTOR__
```
The include guard should be FRAME_DETECTOR_H_

Miceuz replied:
Done

jkff commented on common/libs/framing/soft_rx_buffer.h:
```diff
@@ -0,0 +1,46 @@
+template <int RX_BYTES_MAX> class SoftRxBuffer {
+  uint8_t rx_buf_[RX_BYTES_MAX];
+  uint32_t rx_i_ = 0;
+  RxListener *rx_listener;
```
rx_listener_

Miceuz replied:
Done

jkff commented on common/libs/framing/soft_rx_buffer.h:
```diff
@@ -0,0 +1,46 @@
+    }
+    if (rx_i_ >= RX_BYTES_MAX) {
+      if (rx_listener) {
+        rx_listener->onRxComplete();
```
* This can never be `>`, right?
* As written, one PutByte can cause both OnCharacterMatch AND OnRxComplete, if the marker arrives exactly at RX_BYTES_MAX - is this intentional? If not, then line 38 should be a simple "else".

jkff commented on controller/lib/hal/rx_buf_uart_dma.h:
```diff
@@ -0,0 +1,60 @@
+
+#ifdef TEST_MODE
+  // Puts a byte to rx_buf
+  void test_PutByte(uint8_t b);
```
Why do we need this if we already have a SoftRxBuffer?

jkff replied:
Please address this

Miceuz replied:
This is used in Comms test in later PR. That is more of integration test, but I think there is a benefit of having it to test the whole thing involving frame detection and decoding.

jkff commented on common/libs/framing/soft_rx_buffer.h:
```diff
@@ -0,0 +1,46 @@
+#include "serial_listeners.h"
+#include <stdint.h>
+
+template <int RX_BYTES_MAX> class SoftRxBuffer {
```
Is this class only for testing? If yes, please say so and maybe even #ifdef it on TEST_MODE

Miceuz replied:
No, it's also used for rx from the GUI side

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,144 @@
+
+SoftRxBuffer<LEN> rx_buf('.');
+
+TEST(FrameDetector, TrivialRx) {
```
These tests need some more explanation of what situations they are simulating and why the assertions are what they are.

jkff replied:
Please address this

Miceuz replied:
Well, it was a trivial test, now it does not make sense as we have fakeRx. Test removed.

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,144 @@
+
+constexpr uint8_t MARK = static_cast<uint8_t>('.');
+
+SoftRxBuffer<LEN> rx_buf('.');
```
Use MARK directly?

jkff replied:
This wasn't resolved

Miceuz replied:
It's easier and more clear to the eye writing test cases like `".aaa."` instead of `"\xe2aaaa\xe2"`

Miceuz replied:
Doh! Fixed

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,144 @@
+
+constexpr uint32_t LEN = 20;
+
+constexpr FrameDetector<SoftRxBuffer<LEN>, LEN>::State STATE_LOST =
```
You can simply say `using State = FrameDetector<SoftRxBuffer<LEN>, LEN>::State;`

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,144 @@
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(STATE_WAIT_START, frame_detector.get_state());
```
E.g. why is this WAIT_START - is it because the marker after a LOST is interpreted as an end of frame marker rather than start of frame, and the preceding frame is discarded?

Miceuz replied:
Yes, marker in the LOST condition is interpreted as an END marker.

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,144 @@
+  ASSERT_EQ(STATE_WAIT_START, frame_detector.get_state());
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  rx_buf.PutByte(MARK);
```
Do we always delimit frames with two markers?

Miceuz replied:
Yes, in a way - END marker from previous frame and START marker of the new frame.

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,144 @@
+
+  rx_buf.PutByte(MARK);
+  ASSERT_EQ(STATE_WAIT_START, frame_detector.get_state());
+  EXPECT_TRUE(frame_detector.is_frame_available());
```
This is currently the only test that verifies successful reception of a frame - we need more tests of that, e.g.:
* ability to recover from all kinds of errors
* ability to receive multiple frames
* ability to receive some frames, have an error, and receive more frames
* ability to receive a frame of length almost LEN

I mean, currently the test seems to be written essentially verifying the implementation almost line-by-line, but we need tests that verify the important user-facing properties.

Is it possible e.g. to write a utility function that PutByte's a given std::string, and returns an std::vector<std::string> of received frames, and write a bunch more tests in terms of invocations of this function?

E.g. EXPECT_THAT(Unframe("foo..bar.moo...blah."), testing::ElementsAre("foo", "bar", "moo", "blah"))?

Miceuz replied:
I have started working on that, see updated tests. Will try to implement as you describe.

jkff reviewed (commented):
Thanks! Please edit the PR title and description to be more descriptive.

Miceuz committed:
Update singly byte processing comment; simplify code splitting uint32_t to single bytes

Miceuz committed:
Simplify RxError naming

Miceuz committed:
Return from ISR if an error was detected, don't analyze isCharacterMatch

Miceuz committed:
Rename to adhere to naming standard

Miceuz committed:
Rebase renaming. Also some renaming due to code review

Miceuz committed:
Add FrameDetector documentation

Miceuz committed:
Add more tests

Miceuz committed:
reset frame_available flag on Begin, add more comments

jlebar committed:
Highlight the y=0 gridline.

Makes it easier to see where zero is.

GPC: black-gridline

jlebar committed:
Units: Add missing `const` modifiers.

Some member functions weren't marked as const.

Also remove some unnecessary `inline`s.

GPC: units-const

Miceuz committed:
Refactor FrameDetector to avoid side effects when the start marker was discarded.

Now we keep track of how many markers were received and restart RX only in cases we don't care about the contents.

Miceuz committed:
Add more testing infrastructure

paulovap committed:
[GUI] Update mode selection popup to new design

paulovap committed:
[GUI] Update design of the graphs

Update the design of the graphs to follow new recommendations.

The current implementation of the ScopeView has a very high CPU usage,
making it unable to hold updates at 60Hz. (Actually I can at 90%
CPU usage).

The main reason is the fact we are using two Charts, one To draw a line,
and the other to draw the area. Hopefully, with enough time we can
customize the graphs to make it more efficient.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,206 @@
+ * LOST - FrameDetector is lost in the stream of incoming bytes and will
+ * interpret the next MARK byte as an end of the frame.
+ * WAIT_FOR_START_MARKER - FrameDetector is synced and is waiting for MARK byte
+ * that will denote the start of the frame RECEIVING_FRAME - FrameDetector is
```
Need a line break before RECEIVING_FRAME

Miceuz replied:
Done

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,206 @@
+
+    switch (state_) {
+    case State::LOST:
+      if (IsDataReceived()) {
```
All the states dispatch between IsDataReceived/IsOnlyMarkers/else - so these are essentially a subclassification of OnCharacterMatch event. WDYT about restructuring the method a little like this:

```
void OnCharacterMatch() override {
  if (IsDataReceived()) OnCharacterMatchWithData();
  else if (IsOnlyMarkersReceived()) OnSingleMarkerReceived();
  else { state = LOST; RestartRX(); }
}
void OnCharacterMatchWithData() {
  switch(state)...
}
void OnSingleMarkerReceived() {
  switch(state) ...
}

Miceuz replied:
Yes, I have noticed that also during my last session. Did as you suggested, but not sure if this results in more readable code.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,206 @@
+        // repeated marker char received
+        // this means we have received a 0 lenght frame
+        // ignore it and continue receiving frame bytes
+        // rx_buffer_.RestartRX(this);
```
Should this code be uncommented? If the code should be here, that means there was a bug, please add a test; otherwise delete the code.

Miceuz replied:
Sorry for this mishap. It should be uncommended. Fixed now.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,206 @@
+  void OnRxError([[maybe_unused]] RxError e) override {
+    switch (state_) {
+    case State::LOST:
+    case State::WAIT_FOR_START_MARKER:
```
Is there any harm in just always going into LOST on error?

Miceuz replied:
No, there is no harm. Especially as we don't have timeout error anymore. Frame is truly lost and we will resync on the next frame.

jkff commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,206 @@
+  bool frame_available_ = false;
+  uint8_t frame_buf_[FRAME_BUF_LEN];
+  uint32_t frame_buf_length_ = 0;
+  uint32_t marker_count_ = 0;
```
What are the possible values of this? From mentally tracing the code I'm having trouble understanding whether it ever grows above 1 or 2. If it can take only a couple of values, then it's better to replace it with an enum or something, to emphasize that the state space of the current class is limited.

Miceuz replied:
Hm, I have put a bit of thinking into this today. In normal flow of things it does not go over 2, so yes, your suggestion makes total sense. 
But then I will have to deal with a case that "will never happen" if it's increasing over 2 instead of `marker_count_++`. Now in the current implementation we don't care if it becomes >2 as ProcessData will take care of that correctly. 

Do you think, should we go for it in this light? Do you have a suggestion on what how to implement it in elegant way? Override `++` operator?

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,236 @@
+  EXPECT_FALSE(frame_detector.is_frame_available());
+}
+
+vector<string> fakeRx(const char *frame, FrameDetectorTest &frame_detector) {
```
Change "frame" to std::string, then you can iterate below as "for (char c : frame)"

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,236 @@
+    rx_buf.PutByte(frame[i]);
+    if (frame_detector.is_frame_available()) {
+      string s;
+      s.assign(reinterpret_cast<char *>(frame_detector.TakeFrame()),
```
This should be split in two statements - TakeFrame() is a destructive operation and one can imagine get_frame_length() failing if you call it when the frame is already unavailable.

Miceuz replied:
done

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,236 @@
+  EXPECT_FALSE(frame_detector.is_frame_available());
+
+  vector<string> results = fakeRx(".aaa.", frame_detector);
+  EXPECT_EQ(static_cast<unsigned int>(1), results.size());
```
You can abbreviate this to a one-liner:
EXPECT_THAT(FakeRx(...), testing::ElementsAre("aaa"))

Same below.
You need gunit/gmock.h (or maybe gmock/gmock.h, don't remember)

Miceuz replied:
Thanks, done

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,236 @@
+  EXPECT_EQ(static_cast<unsigned int>(1), results.size());
+  EXPECT_EQ(0, results[0].compare("aaa"));
+
+  EXPECT_TRUE(frame_detector.Begin());
```
In all of these sequences you're basically resetting the FrameDetector. Seems like you could make things much easier by changing FakeRx to create a fresh FrameDetector internally. Then all of these sequences of lines, e.g. 209-214, 216-221 etc., each become a single line.

Miceuz replied:
Getting there, sorry for all the copy paste from the last time, I had very little time and did things hastily.

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,236 @@
+  EXPECT_EQ(static_cast<unsigned int>(1), results.size());
+  EXPECT_EQ(0, results[0].compare("aaa"));
+
+  EXPECT_TRUE(frame_detector.Begin());
```
This seems identical to the previous block

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,236 @@
+  EXPECT_FALSE(frame_detector.is_frame_available());
+  results = fakeRx(".aaaaaaaaaaaaaaaaaaaa..aaa.", frame_detector);
+  EXPECT_EQ(static_cast<unsigned int>(1), results.size());
+  EXPECT_EQ(0, results[0].compare("aaa"));
```
We need some more tests. Currently all these FakeRx tests verify the case when there's only one frame received. How about: more than 1 frame received; 0 frames received; a bunch of data received and some of it is valid frames and some isn't; etc.? This will be very easy to test once you condense these things into one-liners per above.

Miceuz replied:
I have added more tests as per your suggestion, but here I need a little help from you - what if you write some test strings and I write tests for them? It's hard to test own code, when you know how it should work.

jkff reviewed (changes requested):
Thanks Albertas, this is starting to look really nice!

tapemaster committed:
Add sample test data for inspiratory effort detection

jlebar committed:
gitignore .pytype/

GPC: gitignore-pytype

Miceuz committed:
More comments, fix OVERRUN erro name, use nullptr instead of 0

Miceuz committed:
Some renaming and refactoring

Miceuz committed:
Refactor to remove dublicatad code

paulovap committed:
Merge pull request #450 from RespiraWorks/ppinheiro/new_design

Update GUI elements to new design recommendation

jlebar committed:
Merge pull request #459 from RespiraWorks/jlebar/gitignore-pytype

gitignore .pytype/

jlebar committed:
Merge pull request #460 from RespiraWorks/jlebar/black-gridline

Highlight the y=0 gridline.

jlebar committed:
Merge pull request #461 from RespiraWorks/jlebar/units-const

Units: Add missing `const` modifiers.

jlebar committed:
Granularity of Time and Duration 1ms -> 1us.

Motivation: The high-priority loop runs once every few ms, so just a
timer granularity of 1ms is very coarse compared to what's happening in
the system.

GPC: timer-us

jlebar committed:
Build with -wno-error=sign-compare.

This warning is about doing (int)x < (unsigned int)y.

AFAICT this is all noise for us, and in fact clang doesn't even enable
it with -Wall.

GPC: sign-compare

Miceuz committed:
Renaming

Miceuz committed:
Remove trivial test, add more fuzzing tests

jlebar committed:
Remove sample-time checking from PID.

Now PID recomputes every time you call Compute().  Don't want that?
Then don't call compute!

This is made possible by #465, which changes the timer granularity from
1ms to 1us.  Without that change, the clock is too granular to be used
for the high-priority loop, which runs every 10ms -- due to jitter, we
might read 9ms or 10ms when the actual values are more like 9.999ms or
10.0001ms, resulting in an amplification of error.

GPC: PID-no-sample-period

sglow committed:
Implemented PID control of pressure

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,198 @@
+  rx_buf.PutByte(MARK);
+
+  EXPECT_EQ(State::WAIT_FOR_START_MARKER, frame_detector.get_state());
+  ASSERT_TRUE(frame_detector.is_frame_available());
```
Please assert the contents of the frame too

Miceuz replied:
DONE

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,198 @@
+    rx_buf.PutByte(' ');
+  }
+
+  EXPECT_EQ(State::LOST, frame_detector.get_state());
```
Please add a comment why it should be LOST here

Miceuz replied:
DONE

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,198 @@
+  EXPECT_THAT(fakeRx("aaa.bbb.a."), testing::ElementsAre());
+  EXPECT_THAT(fakeRx("aaa.....bbb.a."), testing::ElementsAre("bbb"));
+  EXPECT_THAT(fakeRx(".aaa....bbb.a."), testing::ElementsAre("aaa", "bbb"));
+  EXPECT_THAT(fakeRx("aa.bbb.b...bbbb..s.ss.s...ssss..ss.ssss..aaaaa.sa.aaaa."),
```
Add something starting with ".."?

Miceuz replied:
DONE

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,198 @@
+using FrameDetectorTest = FrameDetector<SoftRxBuffer<LEN>, LEN>;
+constexpr uint8_t MARK = static_cast<uint8_t>('.');
+
+SoftRxBuffer<LEN> rx_buf('.');
```
I think this should be a local variable in each test, otherwise its state is shared between tests and it makes it non-obvious what exactly is being tested, depending on which other tests have run before this one.

Miceuz replied:
DONE

jkff commented on controller/test/frame_detector/frame_detector_test.cpp:
```diff
@@ -0,0 +1,198 @@
+  EXPECT_THAT(fakeRx(".aaa....bbb.a."), testing::ElementsAre("aaa", "bbb"));
+  EXPECT_THAT(fakeRx("aa.bbb.b...bbbb..s.ss.s...ssss..ss.ssss..aaaaa.sa.aaaa."),
+              testing::ElementsAre("bbbb", "s", "ssss", "ss", "aaaaa"));
+}
```
These tests look awesome! I think we need just one more test verifying what I think is the most important property of this protocol - being able to recover from any errors. If we can do that, than we're safe, cause corrupt packets will get caught by CRC.

Take an example string S containing some valid packets - say, "..aaa..bbb..ccc.".
Now we want to verify that, no matter what good or bad stuff we prepend, the result **always** will still contain aaa, bbb, ccc.

We can do that by prepending several (say, up to 10) random events, where each event is either PutByte(marker), or PutByte(non-marker), or an error - i.e. 3 different possible events, 3^10 = 59049 possible prefixes. Also set LEN to a small value (say, 4) so that sometimes several non-markers in a row cover the LEN - 1, LEN - 2, LEN cases. For each of these 59049 prefixes, verify EXPECT_THAT(fakeRx(prefix + S), testing::Contains("aaa", "bbb", "ccc")).

This seems like it should be very simple to implement, but I'm fine if it's in a separate PR.

Miceuz replied:
DONE

jkff reviewed (approved):
Thanks Albertas!

jlebar committed:
Remove unnecessary DbgPrint param from GetResp function.

GPC: dbg-print-param

sglow committed:
Minor changes based on reviewer feedback

sglow committed:
Merge pull request #469 from RespiraWorks/issue_421

Implemented PID control of pressure

tapemaster committed:
Merge pull request #457 from RespiraWorks/inspiratory-attempt-sample-data

Add sample test data for inspiratory effort detection

jlebar committed:
Merge pull request #470 from RespiraWorks/jlebar/sign-compare

Build with -wno-error=sign-compare.

jlebar committed:
Merge pull request #471 from RespiraWorks/jlebar/dbg-print-param

Remove unnecessary DbgPrint param from GetResp function.

jlebar committed:
Merge pull request #462 from RespiraWorks/jlebar/pid-time

Revert "Removed time checks from PID library."

jlebar committed:
Merge pull request #465 from RespiraWorks/jlebar/timer-us

Granularity of Time and Duration 1ms -> 1us.

jlebar committed:
Merge pull request #463 from RespiraWorks/jlebar/tv-zeroing

First pass at TV zeroing.

jlebar committed:
Merge pull request #467 from RespiraWorks/jlebar/PID-no-sample-period

Remove sample-time checking from PID.

paulovap committed:
Merge pull request #464 from RespiraWorks/ppinheiro/update_graphs

[GUI] Update design of mode selection popup and graphs

jlebar committed:
Fix units errors in sensors_test.

In multiple places in sensors_test, we use a tolerance in one unit (e.g.
kPa) to compare two values in a different unit (e.g. cmh2o).

Fix this by introducing EXPECT_{FLOW,PRESSURE,VOLUME}_NEAR macros.

GPC: flow-sensors-test

jlebar committed:
Update to new 3/4" Venturi specifications.

GPC: new-venturi

jlebar committed:
Disable sign-compare by using -Wno.

Rather than disabling it via -Wno-error.  This way it's totally silent.

GPC: sign-compare-take-2

jlebar committed:
Move setpoint_pressure field into ControllerState.

It doesn't belong in ActuatorState because the setpoint is not an
actuator!

GPC: move-setpoint-pressure

jlebar committed:
Merge pull request #476 from RespiraWorks/jlebar/sign-compare-take-2

Disable sign-compare by using -Wno.

jlebar committed:
Merge pull request #474 from RespiraWorks/jlebar/flow-sensors-test

Fix units errors in sensors_test.

jlebar committed:
Move flow zeroing into TVIntegrator.

GPC: flow-zeroing

jlebar committed:
Use nullptr instead of 0 for a void* param.

GPC: nullptr

jlebar committed:
Sleep for a few seconds at startup.

Due to a hardware issue, the blower always spins up when the board
restarts.  Wait for the blower to spin down, otherwise the sensors will
be miscalibrated.

GPC: sleep-at-startup

sglow committed:
Adjusted PID tuning to make system more stable.

Also reset PID state when ventilator is off

jlebar committed:
Merge pull request #481 from RespiraWorks/jlebar/move-setpoint-pressure

Move setpoint_pressure field into ControllerState.

jlebar committed:
Merge pull request #482 from RespiraWorks/jlebar/sleep-at-startup

Sleep for a few seconds at startup.

jlebar committed:
Merge pull request #480 from RespiraWorks/jlebar/nullptr

Use nullptr instead of 0 for a void* param.

jlebar committed:
Move blower_fsm implementations into cpp file.

Main motivation is to allow us to declare constants and functions that
are "private" to the file.

GPC: move-fsm-impls

jlebar committed:
Rise from PIP to PEEP in 100ms.

Instead of going directly from PIP to PEEP, transition over 100ms.

Citation: https://respiraworks.slack.com/archives/C011CJQV4Q7/p1591763842312500?thread_ts=1591759016.310200&cid=C011CJQV4Q7

GPC: pip-rise-time

jlebar committed:
Units: Allow x / y to return a raw ratio.

In a future PR, I want to divide two Durations.

Also mark more operators as const, and test that they are in fact const.

Please forgive the unironic use of pointer-to-member-function.

GPC: units-ratio

Miceuz commented on common/libs/framing/frame_detector.h:
```diff
@@ -0,0 +1,212 @@
+  FrameDetector(RxBuffer &t) : rx_buffer_(t){};
+
+  // Starts frame detector
+  [[nodiscard]] bool Begin() {
```
nope

Miceuz committed:
Create new fx_buffer for each test. Add more tests.

sglow committed:
Added automatic probing of stepper chips

This also reverses the direction of motion of the pinch valves
to conform to the standard wiring diagrams.

jlebar committed:
Merge pull request #475 from RespiraWorks/jlebar/new-venturi

Update to new 3/4" Venturi specifications.

jlebar committed:
Fix formatting in pid.cpp.

Somehow escaped clang-format.

GPC: pid-formatting

sglow committed:
Minor edits based on review feedback

sglow committed:
Merge pull request #487 from RespiraWorks/autoprobe_steppers

Added automatic probing of stepper chips

sglow committed:
Minor improvements to data tracing

Timestamps are added to saved trace data
controller is queried to find sample time rather then hard code it

sglow committed:
Added support for the controller audio alarm

sglow committed:
Comment changes

sglow committed:
Minor edits based on review feedback

sglow committed:
Merge pull request #490 from RespiraWorks/add_buzzer_support

Added support for the controller audio alarm

sglow committed:
Spelling error

jlebar committed:
Bump gcc version to 9.2.1.

This lets us use designated initializers with fields which have
nontrivial constructors (e.g. std::optional).

This is still not a bleeding-edge gcc, so I think we should be OK.

GPC: gcc9

jlebar committed:
Remove trailing whitespace in sample-data/README.md

GPC: trailing-whitespace

jlebar committed:
Eagerly home pinch valves.

Instead of homing them when you first start ventilation and then
hard-disabling them when you stop (requiring re-homing before we begin
again).

Also fix a TODO in controller.cpp.

Note that the stepper motors consume a nontrivial amount of power in
this mode.

GPC: controller-todo

sglow committed:
Merge pull request #489 from RespiraWorks/trace_improvements

Minor improvements to data tracing

jlebar committed:
Merge pull request #494 from RespiraWorks/jlebar/pid-formatting

Fix formatting in pid.cpp.

jlebar committed:
Merge pull request #495 from RespiraWorks/jlebar/trailing-whitespace

Remove trailing whitespace in sample-data/README.md

sglow committed:
Added test of PID reset

sglow committed:
Merge pull request #484 from RespiraWorks/adjust_tuning

Adjusted PID tuning to make system more stable.

jlebar committed:
Merge pull request #483 from RespiraWorks/jlebar/flow-zeroing

Move flow zeroing into TVIntegrator.

jlebar committed:
Merge pull request #485 from RespiraWorks/jlebar/units-ratio

Units: Allow x / y to return a raw ratio.

jlebar committed:
Merge pull request #486 from RespiraWorks/jlebar/move-fsm-impls

Move blower_fsm implementations into cpp file.

jlebar committed:
Merge pull request #488 from RespiraWorks/jlebar/pip-rise-time

Rise from PIP to PEEP in 100ms.

jlebar committed:
Merge pull request #496 from RespiraWorks/jlebar/gcc9

Bump gcc version to 9.2.1.

jlebar committed:
Merge pull request #497 from RespiraWorks/jlebar/controller-todo

Eagerly home pinch valves

Miceuz committed:
Remove timeout feature from uart_dma driver as we don't really use it

Miceuz committed:
Chase timeout remove

Miceuz committed:
Add infrastructure for error testing

Miceuz committed:
Remove more timeout. Stop RX after error as DMA transmission is configured to stop

Miceuz committed:
Add test for errors during RX

Miceuz committed:
Add ultimate fuzzing test prepending random noise to a valid frame

Miceuz committed:
Add last fixes for review

Miceuz commented:
phew, now I hope this is really done

jkff reviewed (commented):
Thanks Albertas! Code-wise, this looks great, but there's two things:

1) It doesn't pass test.sh on my machine with the following error:

```
...
ar rc .pio/build/native/lib658/libframing.a
ar: no archive members specified
```

This is because framing is a header-only library. There's two ways to fix this: add some fake .cpp file to it that has at least one symbol, or set lib_archive = false (https://docs.platformio.org/en/latest/projectconf/section_env_library.html#lib-archive) - but then I'm hitting other issues.

2) Let us hold off merging this until after the CoVent challenge deadline next week, because, while this is a critical component of the project as a whole, it is not a critical component of the subset we're delivering by then.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RespiraWorks/Ventilator?pullRequest=441) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/RespiraWorks/Ventilator?pullRequest=441) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: Miceuz<br/>:x: jlebar<br/>:x: sglow<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RespiraWorks/Ventilator?pullRequest=441) it.</sub>
